### PR TITLE
Wrap literals around expression refs

### LINKED ISSUE #27185: Postgres "specific dates" filter errors with datetime custom expression functions
When using a Postgres database, applying a "specific dates" filter with a custom expression function results in an error:

`ERROR: operator does not exist: timestamp with time zone >= character varying Hint: No operator matches the given name and argument types. You might need to add explicit type casts.`

image

Steps to reproduce: 
1. Create a new question, selecting `Orders`
2. Create a custom expression with the formula `datetimeAdd([Created At], 1, "month")`
3. Create a filter on the custom expression column, selecting "specific dates"
image
4. Observe the error.

`ERROR: operator does not exist: timestamp with time zone >= character varying Hint: No operator matches the given name and argument types. You might need to add explicit type casts. Position: 1091`

Can replicate with `now`, `datetimeSubtract`, or `convertTimezone` by changing step 2.

- Your databases: Postgres
- Metabase version: 0.45.1

The problem might exist for other DBMS other than Postgres.

### PR DESCRIPTION
**Pull request by snoe**
Fixes #27185

Expression refs need to be handled like field refs as far as type hints in opts are concerned. Otherwise we don't pass the correct parameter types to the drivers.

### CODE REVIEW

**Comment by snoe** [comment]
This requires https://github.com/metabase/metabase/pull/38874 to be merged first

File: test/metabase/query_processor/middleware/wrap_value_literals_test.clj [hunk 1]
```diff
@@ -211,6 +211,13 @@
               {:source-query {:source-table $$checkins
                               :filter       [:> $date "2014-01-01"]}}))))))
 
+(deftest ^:parallel wrap-literals-around-expressions
```
Comment (lbrdnk) [inline test/metabase/query_processor/middleware/wrap_value_literals_test.clj:214]: Should not test names end with `-test`?

File: test/metabase/query_processor/middleware/wrap_value_literals_test.clj [hunk 2]
```diff
@@ -211,6 +211,13 @@
...
+  (testing "does wrapping literals work against expressions? (#27185)"
+    (wrap-value-literals
```
Comment (lbrdnk) [inline test/metabase/query_processor/middleware/wrap_value_literals_test.clj:216]: I assume here we are testing that `qp.wrap-value-literals/wrap-value-literals` does not throw. If my assumption is correct, being explicit about that would make the code easier to reason about.

File: test/metabase/query_processor/middleware/wrap_value_literals_test.clj [hunk 3]
```diff
@@ -211,6 +211,13 @@
...
+      (lib.tu.macros/mbql-query checkins
+        {:expressions {"foo" $date}
+         :filter       [:> [:expression "foo" {:base-type :type/DateTime}] "2014-01-01"]}))))
```
Comment (lbrdnk) [inline test/metabase/query_processor/middleware/wrap_value_literals_test.clj:219]: NIT: whitespace.

**Review by lbrdnk — APPROVED** [review]
LGTM. I've tested the changes rebased on `fix-expression-datetime-binned-filters` branch. Apart from fixing the issue in the PR description, it also fixes the #37165. That's something to be confirmed when this PR and #38874 are in master.

**Review by metamben — APPROVED** [review]

File: test/metabase/query_processor/middleware/wrap_value_literals_test.clj (hunk 2 above)
Comment (snoe) [reply test/metabase/query_processor/middleware/wrap_value_literals_test.clj:216]: Thanks, dumb miss on my end, repl testing and forgot to make an assertion.

**Pull request merged**